### PR TITLE
chore: LSM changes to distribution

### PR DESCRIPTION
**Pull request by mpoke**
## Description

Related: #18200  

Port the LSM changes to the distribution module to 0.47. Also, add the staking changes  necessary to deal with errors. 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] added `!` to the type prefix if API or client breaking change
* [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
* [ ] provided a link to the relevant issue or specification
* [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/build/building-modules/01-intro.md)
* [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
* [ ] added a changelog entry to `CHANGELOG.md`
* [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] updated the relevant documentation or specification
* [ ] reviewed "Files changed" and left comments if necessary
* [ ] run `make lint` and `make test`
* [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] confirmed all author checklist items have been addressed 
* [ ] reviewed state machine logic
* [ ] reviewed API design and naming
* [ ] reviewed documentation is accurate
* [ ] reviewed tests and test coverage
* [ ] manually tested (if applicable)

### CODE REVIEW

**Comment by sainoe** [comment]
Could please add [BeforeTokenizeShareRecordRemoved](https://github.com/cosmos/cosmos-sdk/pull/16747/files#diff-e91f5c93caa2f55898d8b564433aad05f8b45a3803ad3733dd9fa808d265b381R104-R112) to the hook methods of the distribution module?

**Comment by mpoke** [comment]
> Could please add [BeforeTokenizeShareRecordRemoved](https://github.com/cosmos/cosmos-sdk/pull/16747/files#diff-e91f5c93caa2f55898d8b564433aad05f8b45a3803ad3733dd9fa808d265b381R104-R112) to the hook methods of the distribution module?

Nice catch

**Comment by sainoe** [comment]
Okay for merging

File: x/distribution/keeper/delegation_test.go [hunk 1]
```diff
@@ -98,6 +98,10 @@ func TestCalculateRewardsBasic(t *testing.T) {
 	require.Equal(t, sdk.DecCoins{{Denom: sdk.DefaultBondDenom, Amount: math.LegacyNewDec(initial / 2)}}, distrKeeper.GetValidatorAccumulatedCommission(ctx, valAddr).Commission)
 }
 
+func TestWithdrawTokenizeShareRecordReward(t *testing.T) {
+	// TODO add LSM test
```
Comment (MSalopek) [inline x/distribution/keeper/delegation_test.go:102]: Let's merge without this test as I will add it a little later.

**Review by MSalopek — APPROVED** [review]
LGTM!

Missing tests will be added in subsequent PRs.

**Pull request merged**